### DISCUSSION TITLE
Literal attribute not inferred from `__init__` signature — is this a bug?

### DISCUSSION THREAD

**Discussion by dmontagu** [microsoft/pyright#4812] [General]
I discovered an issue today where pyright does not seem to use `Literal` types as the assigned attribute types for instance attributes in a class's `__init__` if the type annotation came from the `__init__` signature. (I am used to this behavior from mypy.)

More concretely:

```python
from typing import Literal

FruitKind = Literal['apple', 'banana']
class MyClass:
    def __init__(self, fruit: FruitKind):
        self.fruit = fruit

# reveal_type(MyClass(fruit='apple').fruit)
# Pyright says: Type of "MyClass(fruit='apple').fruit" is "str"
# Mypy says: Revealed type is "Union[Literal['apple'], Literal['banana']]"

def f(fruit: FruitKind) -> None:
    pass


f(MyClass(fruit='apple').fruit)  # pyright gives an error; mypy doesn't
```

I'll note that explicitly setting a type annotation on the attribute fixes the type errors (i.e., `self.fruit: FruitKind = fruit`).

Is this behavior intentional? I.e., should I be explicitly providing a type attribute on the annotation? Or should this be considered a bug in pyright?

**Comment by erictraut** [comment]
Yes, this behavior is intentional. You haven't provided a type declaration for the instance variable `MyClass.fruit`, so its type is inferred. The inferred type of a variable does not include literal types. If your intent is for this instance variable to be limited to literal types, you need to declare it as such.

If you assigned `self.fruit` a value of `0` in the constructor, you almost certainly wouldn't want pyright to infer the type of `MyClass.fruit` to be `Literal[0]`.

**Reply by dmontagu** [reply]
Well, I assigned it a variable (`fruit`) that had type `Fruit` — it's surprising to me that the annotated type from the `__init__` signature is getting replaced with `str`. Are there any cases besides `Literal` where this happens?

If this isn't going to change, that's fine, but I'll also note that PyCharm's behavior matches mypy on this:
<img width="478" alt="image" src="https://user-images.githubusercontent.com/35119617/226495113-eb8eec2e-8572-40fc-9638-c6e467e096fb.png">

I suppose really my question is:
> The inferred type of a variable does not include literal types

Is this a consequence of an implementation detail, or is the demonstrated behavior considered preferable?

**Reply by erictraut** [reply]
Yes, you can assume this won't change. Pyright's behavior is consistent whereas mypy (and apparently PyCharm) are not. [This documentation](https://microsoft.github.io/pyright/#/mypy-comparison?id=inference-of-list-set-and-dict-expressions) explains the inconsistency.

**Reply by dmontagu** [reply]
Also,

> If you assigned self.fruit a value of 0 in the constructor, you almost certainly wouldn't want pyright to infer the type of MyClass.fruit to be Literal[0].

~I would say that if I had _annotated_ it as being of type `Literal[0]`, I would still expect the attribute to be of type `Literal[0]` (ignoring the assigned value). I could see this being relevant, for example, in cases of discriminated unions where I want the `__init__` function should accept the discriminator value (I'm thinking something like `if data['kind'] == 0: return Case0(**data)`). Generally if I didn't want this behavior, I wouldn't annotate the `__init__` signature in that way.~

At any rate, I appreciate the response; we'll move forward just being more explicit with the annotations for literal-valued attributes.

-----

Edit: I initially misinterpreted your comment as saying `def __init__(fruit=0): self.fruit = fruit` should not result in `self.fruit` having type (`Literal[0]`), with I agree with but didn't see the connection. I now realize you meant `def __init__(): self.fruit = 0`, and I see the inconsistency in handling of `Literal[0]` that you are describing.

**Reply by dmontagu** [reply]
>Yes, you can assume this won't change. Pyright's behavior is consistent whereas mypy (and apparently PyCharm) are not. [This documentation](https://microsoft.github.io/pyright/#/mypy-comparison?id=inference-of-list-set-and-dict-expressions) explains the inconsistency.

GitHub didn't show me this response until I shared my follow-up 🤦 thank you very much for sharing! This is what I was looking for.
